fix duplicate email invite form error, adds tests

clarktr1 opened the pull request:
Closes #69 and #70 

@ldko @somexpert This is ready for review.

Checking duplicate emails in Invites/Users
- This was a simple change. A function that checks the InviteItem against current Invites and returns if they have emails that are the same. A test was written for this.
- The error for this is rendered whenever the 'email' field in the form is changed. This should get displayed over the form in question.

Normalizes emails
- The `clean_email` function was overridden to strip whitespace and lowercase the value.
- This can be tested by creating a new invitation and looking at the email in the UI.

Check duplicate emails in current Invite form
- This issue sprang up from trying to resolve the first email issue. If there were duplicate emails in more than one row, Django would accept all of the rows regardless and there would be duplicate emails across several Invitations. 
- A formset was added to validate the emails in the form were all unique. 
- A new partial was added to the `invite.html` to render the error/s from the form. This new error div renders at the top of the form with the email in question included. This does not disable the submit button like the eventlistener validations do. This is because this particular check only happens server side rather than client side. 

To test these changes
- Create a Superuser if one doesn't exist with `./manage.py createsuperuser`.
- Try to create an Invitation with the same email as the superuser (should raise an error)
- Create an Invitation with a different value
- Try to create an Invitation with the same email as the just created Invitation (should raise an error).
- Create 2+ rows where 2+ have duplicate emails (should raise an error after the form is submitted).

clarktr1 committed:
fix duplicate email invite form error, adds tests

clarktr1 committed:
remove debug statements

ldko commented on invite/forms.py:
```diff
@@ -10,20 +10,21 @@ def validate_username(value):
 
-
 def validate_user_email_exists(value):
-    if value in User.objects.all().values_list('email', flat=True):
+    if value.lower() in User.objects.all().values_list('email', flat=True):
         raise ValidationError('The email provided already belongs to a user')
 
+def validate_invitation_email_exists(value):
+    if value.lower() in Invitation.objects.all().values_list('email', flat=True):
+        raise ValidationError(f'{value} already belongs to a user or pending invitation')
+
```
What do you think about combining these into one function, since we want to do these checks in tandem anyway?

clarktr1 replied:
We can do that. Something like `validate_email_exists`.

ldko commented on invite/forms.py:
```diff
@@ -115,6 +117,19 @@ def clean(self, *args, **kwargs):
 
+class BaseInviteItemFormSet(forms.BaseFormSet):
+    def clean(self):
+        """
+        Validates that there are no duplicate email addresses across forms.
+        """
```
```suggestion
        """Validate there are no duplicate email addresses across forms."""
```

ldko replied:
I was also suggesting that this should [all be on one line](https://peps.python.org/pep-0257/#one-line-docstrings).

ldko commented on invite/forms.py:
```diff
@@ -10,20 +10,21 @@ def validate_username(value):
-
 def validate_user_email_exists(value):
-    if value in User.objects.all().values_list('email', flat=True):
+    if value.lower() in User.objects.all().values_list('email', flat=True):
```
I wonder if we need to do something like in the above `validate_user_email` function with lowercasing emails that we pull from the User table because those can be created outside of django-invite and may not be guaranteed to be lowercase. The part like:
```
    insensitive_emails = (
        [e.lower() for e in User.objects.all().values_list('email', flat=True)]
    )
```

ldko commented on invite/forms.py:
```diff
@@ -115,6 +117,19 @@ def clean(self, *args, **kwargs):
+        emails = []
+        for form in self.forms:
+            email = form.cleaned_data.get('email')
+            if email:
+                if email in emails:
+                    raise ValidationError(f"{email} already belongs to another invitation in the current form")
```
Could combine the conditions and use single quotes for consistency.
```suggestion
            if email and email in emails:
                raise ValidationError(f'{email} already belongs to another invitation in the current form')
```

ldko commented on invite/forms.py:
```diff
@@ -133,7 +148,7 @@ class InviteItemForm(forms.ModelForm):
     )
     email = forms.EmailField(
-        validators=[validators.validate_email, validate_user_email_exists],
+        validators=[validators.validate_email, validate_user_email_exists, validate_invitation_email_exists],
```
Do we need `validators.validate_email` here or is it in the default validators?

clarktr1 replied:
I don't think we need it here either. Django provides the email validation for us as the default validator.

ldko commented on invite/forms.py:
```diff
@@ -144,6 +159,12 @@ class InviteItemForm(forms.ModelForm):
         ),
     )
 
+    def clean_email(self):
+        email = self.cleaned_data.get('email')
+        if email:
+            email = email.strip().lower()
+        return email
```
Where does this get called?

clarktr1 replied:
`clean_email` is an overridden method that gets called inside of `clean()` during the process of creating the `cleaned_data` dictionary. [Django docs](https://docs.djangoproject.com/en/5.1/ref/forms/validation/#form-and-field-validation)

clarktr1 commented on invite/forms.py:
```diff
@@ -51,6 +52,7 @@ class SignupForm(forms.Form):
         ),
     )
     email = forms.EmailField(
+        validators=[validators.validate_email],
```
Does not look like we need this here.

clarktr1 commented on invite/forms.py:
```diff
@@ -10,20 +10,21 @@ def validate_username(value):
 
-
 def validate_user_email(value):
     insensitive_emails = (
         [e.lower() for e in User.objects.all().values_list('email', flat=True)]
     )
-    assert False, insensitive_emails
     if value.lower() not in insensitive_emails:
         raise ValidationError('The email provided doesn\'t belong to any user')
```
Not explicitly. Looking at it, it's definitely supposed to be used [here](https://github.com/unt-libraries/django-invite/blob/master/invite/forms.py#L239)

ldko replied:
We can delete `validate_user_email`, since it is not being used.

ldko replied:
OK, backpedaling, if it works to use `validate_user_email` in the `email` field `validators`of the `IForgotForm` as a replacement for the `clean_email` method that is defined in the class, I think we should do that :D.

clarktr1 commented:
https://github.com/unt-libraries/django-invite/blob/280242c6b9283c069459f0819e8c0ccc33ab73a7/invite/views.py#L343

Remove this debug statement. 👎

clarktr1 reviewed (commented):
https://github.com/unt-libraries/django-invite/blob/280242c6b9283c069459f0819e8c0ccc33ab73a7/invite/views.py#L343
Remove this 👎

ldko reviewed (commented):
I had a chance to look at/run the tests and run the app with the changes and have a few more comments.

clarktr1 commented on invite/templates/invite/invite.html:
```diff
@@ -197,6 +200,15 @@
     {% if perms.invite.add_invitation %}
         <h3>Whom would you like to invite, {{ user|title }}?</h3>
+        {% if errors %}
+        <div class="alert">
```
I was wondering why I didn't add this the first time around. The function that validates fields and adds the error messages looks specifically for the 'alert-error' class to not be present in any divs before the submit button in enabled again. Since the error in this div is checked on the form submission, there would be no way to resubmit the form to remove the error if the container still exists. I can make it the same using inline styling, though!

```
 <div class="alert" style="background-color: #f2dede; color: #b94a48; border-color: #eed3d7;">
            <ul>
                {% for error in errors %}
                        <li>{{ error }}</li>
                {% endfor %}
            </ul>
        </div>
```

clarktr1 committed:
add user dupe checking, remove redundant validations, modify dupe  tests

clarktr1 committed:
flake8

clarktr1 commented:
@ldko @somexpert This is ready for another pass.
# Changes
* combined the email validation functions since they occur at the same time. 
* changed the way to grab insensitive emails to just use `filter(<field_name>__iexact=value).exists()` to make sure emails can we be checked in case of failure normalizing as well as letting Django handle the checks.
* clean() handles email and user validation. 
* removed redundant validators
* made errors specific to the exception they're raising
* The new error div for has inline styling instead of sharing the `alert-error` class (from Bootstrap 2.x.x). This is due to how the frontend functions detect, remove, and enable the submission button again.
* Added some small checks into test to validate clean, clean_email, and username dupe validation

Behaviors that should occur
1) No email that already exists as a User email or Invitation email can be added to the form. If so, submission button locks and errors appear over forms in question.
2) If the User tries to submit duplicate emails/users across multiple forms, they will be redirected to the form again with all errors relating to the submission. Submission button stays open and the error stays present until next submission
* This also works in the event of different duplicates across multiple forms.

ldko commented on tests/test_all.py:
```diff
@@ -686,6 +742,25 @@ def test_check_returns_false_with_no_args(self):
+            {
+                'form-MAX_NUM_FORMS': [''],
+                'form-INITIAL_FORMS': ['0'],
+                'form-TOTAL_FORMS': ['2'],
```
Likely of no consequence, but for appearances, this could be
```suggestion
                'form-TOTAL_FORMS': ['1'],
```

ldko commented on invite/forms.py:
```diff
@@ -12,17 +11,15 @@ def validate_username(value):
-    assert False, insensitive_emails
-    if value.lower() not in insensitive_emails:
-        raise ValidationError('The email provided doesn\'t belong to any user')
+    if User.objects.filter(email__iexact=value).exists():
```
If we keep this to use in `validataors` of the `IForgotForm`, I think this needs to change to:
```suggestion
    if not User.objects.filter(email__iexact=value).exists():
```

ldko commented on tests/test_all.py:
```diff
@@ -344,7 +344,8 @@ def test_signup_submit_same_email(self):
 
-    def test_invite_submit_same_email(self):
+    def test_invite_user_duplicate_email(self):
+        """" Test for Inivtation email matching current User email """
```
Looking through the preexisting files in this repo, I see there is not a consistent inclusion of docstrings and syntax varies, but I recommend triple quotes (note 4 opening quotes here), no space following or leading into the triple quotes of the docstring, and use a period (also note spelling typo).
```suggestion
        """Test for Invitation email matching current User email."""
```

clarktr1 committed:
normalize docstrings in tests, add validation to iforgot and remove clean_email override

clarktr1 commented:
@somexpert @ldko This is ready for another review when you have a moment

somexpert reviewed (commented):
So in addition to the few small comments ~above~ below, I am wondering if we should also adjust `validate_username()` to check pending invitations. Currently I am able to duplicate usernames for invitations in different forms because the username is only checked against existing users.

clarktr1 committed:
update tests, error messages, check username for Invitations

clarktr1 commented:
@somexpert Addressed your changes whenever you have a moment

somexpert reviewed (approved):
Looks good to me! :thumbsup:

clarktr1 committed:
split username validation for signup form and invite form

clarktr1 commented:
@ldko @somexpert I've split the checks into separate functions and just made the invite form have 2 validators.

ldko reviewed (approved):
This is working for me. Thanks, @clarktr1 ! if you want to add yourself to the contributors in the README, please do!

clarktr1 committed:
update README contributors
